File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/ordering/SendOrderConfirmation.kt
package nl.zappymart.application.ordering

import nl.zappymart.application.ports.CustomerRepository
import nl.zappymart.application.ports.Mailer
import nl.zappymart.application.ports.OrderConfirmation
import nl.zappymart.application.ports.OrderRepository
import nl.zappymart.domain.ordering.OrderPlaced

class SendOrderConfirmation(
    private val customers: CustomerRepository,
    private val orders: OrderRepository,
    private val mailer: Mailer,
) {

    fun handle(event: OrderPlaced) {
        val customer = customers.findById(event.customerId) ?: return
        val order = orders.findForCustomer(event.orderId, event.customerId) ?: return
        mailer.send(OrderConfirmation(customer.email, customer.name, order.number, order.total))
    }
}

File: backends/dotnet/src/Zappy.Application/Ordering/SendOrderConfirmation.cs
using Zappy.Domain;

namespace Zappy.Application;

public sealed class SendOrderConfirmation(
    IOrderRepository orders,
    ICustomerRepository customers,
    IMailer mailer) : IDomainEventHandler<OrderPlaced>
{
    public async Task Handle(OrderPlaced orderPlaced, CancellationToken cancellationToken)
    {
        var customer = await customers.WithId(orderPlaced.CustomerId, cancellationToken);
        if (customer is null)
        {
            return;
        }

        var order = await orders.OfCustomerWithId(orderPlaced.CustomerId, orderPlaced.OrderId, cancellationToken);
        if (order is null)
        {
            return;
        }

        await mailer.SendOrderConfirmation(customer.Email, customer.Name, order, cancellationToken);
    }
}
